Update prometheus operator 0.43.2

ipochi opened the pull request:
This PR updates the Prometheus operator component to the latest upstream chart that is now renamed from `prometheus-operator` to `kube-prometheus-stack`.

Release notes: 


Steps(1-4) must be performed before updating the cluster i.e before running `lokoctl cluster apply`. 
Steps 5,6 must be performed after updating the cluster ie after running `lokoctl cluster apply`.

Upgrading this component requires downtime and some steps to be executed manually before upgrading:

1. Patch the PersistenceVolume created/used by the prometheus-operator chart to `Retain` claim policy:

    ```console
    kubectl patch pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-prometheus-0")].metadata.name}')

    kubectl patch pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-alertmanager-0")].metadata.name}')
    ```

    **Note:** To execute the above command, the user must have a cluster wide permission.

2. Uninstall the **prometheus-operator** release and delete the existing PersistentVolumeClaim, and verify PV become Released.

    ```console
    lokoctl component delete prometheus-operator
    kubectl delete pvc/data-prometheus-prometheus-operator-prometheus-0" -n monitoring
    kubectl delete pvc/data-alertmanager-prometheus-operator-alertmanager-0" -n monitoring
    ```

    You can choose to remove all your existing CRDs (ServiceMonitors, Podmonitors, etc.) if you want to.

3. Remove current `spec.claimRef` values to change the PV's status from Released to Available.
    ```console
    kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-prometheus-0")].metadata.name}')

    kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-alertmanager-0")].metadata.name}')
    ```
    **Note:** To execute the above command, the user must have a cluster wide permission.

4. Make sure that the `storage_class` and `prometheus.storage_size` are unchanged during the upgrade process.

5. If monitoring was enabled for rook, contour, metallb components, then the user needs to update the said components as well. Backwards compatibility exists and its not mandatory to update the said components before installing updated prometheus-operator component.

6. Proceed to a fresh **prometheus-operator** component installation. The new release should now re-attach your previously released PV with its content.

Closes #1079

invidian commented on assets/charts/components/prometheus-operator/charts/grafana/Chart.yaml:
```diff
@@ -11,7 +11,9 @@ maintainers:
 sources:
 - https://github.com/grafana/grafana
-version: 5.7.10
+version: 5.8.16
```
Why do we need to update it individually? Shouldn't this be done by upstream?

invidian reviewed (commented):
Just one question, otherwise LGTM.

invidian reviewed (approved):
LGTM

In 5. of release notes, I would add "... otherwise metrics from those components won't be collected."

ipochi commented:
> LGTM
> 
> In 5. of release notes, I would add "... otherwise metrics from those components won't be collected."

done.

surajssd commented on assets/charts/components/prometheus-operator/CONTRIBUTING.md:
```diff
@@ -1,10 +1,12 @@
 # Contributing Guidelines
```
In this first commit, can you provide the version of chart and app for each sub chart? Also the commit says update prometheus-operator but we don't update that.

surajssd commented on assets/charts/control-plane/kubernetes/templates/etcd-endpoints.yaml:
```diff
@@ -7,7 +7,7 @@ metadata:
     # These labels are to make sure that the Prometheus Operator ServiceMonitor picks this
     # endpoints correctly.
-    app: prometheus-operator-kube-etcd
+    app: kube-prometheus-stack-kube-etcd
```
This mandates the user to update the prometheus-operator immediately after cluster upgrade. And since prometheus-operator is updated the user has to update their entire infra along with components like contour, rook & metallb.

We should provide older service and endpoint along with newer one. So user can take their time to do similar changes in their infra and then decide to upgrade prometheus-operator when they need.

ipochi replied:
@surajssd Ideally we should not advocate to the users to stop only at cluster update. A lokoctl binary should be consumed in full i.e cluster update as well as the update of components the user has installed.

What if the user is comfortable using the the component installed in previous versions and Lokomotive has moved forward 2 or more releases. A must note in our cluster upgrade process should be this "ensure all components are updated per the release notes(which should mention the components that were updated) before starting the next update process"

@invidian @iaguis Your thoughts on this ?

invidian replied:
>We should provide older service and endpoint along with newer one. So user can take their time to do similar changes in their infra and then decide to upgrade prometheus-operator when they need.

IMO we do not support partial upgrades and I doubt we are ready to do so. I think it's not a big deal to roll the update across all the components.

surajssd replied:
@invidian what about the user's applications?

surajssd replied:
> I think it's not a big deal to roll the update across all the components.

Not true for a prod cluster serving traffic.

invidian replied:
>@invidian what about the user's applications?

Sounds like out of scope for me. Also, we discussed in the past that this Prometheus should be treated as a "system" one.

>Not true for a prod cluster serving traffic.

I'm not saying they should be all updated at once, but until all of them are updated, I'd treat the cluster as "in maintenance window". And if something goes wrong with the update, updated components should be rolled back to previous versions to remain consistent (though we don't test that).

iaguis replied:
I agree we shouldn't advocate being in a partially updated state and we shouldn't support that in general. However, we should make updates easy for users to apply and I think it's reasonable for users running in production to take some time to complete an update, so if we can make this easier and don't require them to update all their infra right after the control plane is update to not have a monitoring interruption we should do that.

I don't know the details in this case so feel free to ignore me if this is nonsense but maybe we can expose both endpoints here and remove the old one in the next release?

johananl replied:
>This mandates the user to update the prometheus-operator immediately after cluster upgrade. And since prometheus-operator is updated the user has to update their entire infra along with components like contour, rook & metallb.

- As far as I can tell upgrades should be atomic as much as possible. I actually *expect* all components which have newer versions in a release to be upgraded in a single operation. Why is it a problem to upgrade prometheus-operator, Contour, Rook and MetalLB "in one go"?
- What do you mean by "their entire infra" @surajssd? Can you name the specific things which need to be updated?
- As a user, what is the implication if the first approach is taken vs. the second approach?

johananl replied:
Another thought: If upgrading all components at once would have unreasonable implications from the user's perspective *and* there is nothing which can be done manually during the upgrade to mitigate these implications, another option is to release a backward-compatible update and do a "cleanup" in a later update.

surajssd replied:
@johananl that's what I am proposing doing a backward compatible update and clean up later.

surajssd replied:
> What do you mean by "their entire infra" @surajssd? Can you name the specific things which need to be updated?

Viz. ServiceMonitors and PrometheusRules objects across the cluster will need an update.

ipochi replied:
This is because label `app: prometheus-operator` is now changed to `app: kube-prometheus-stack`.

This change would also escalate into any user specific applications that are being monitored.

invidian replied:
>This change would also escalate into any user specific applications that are being monitored and consequentially face downtime in collecting metrics.

User applications shouldn't use those labels. As far as I know, we treat shipped Prometheus instance as a system-one, integrated with components.

ipochi replied:
> > This change would also escalate into any user specific applications that are being monitored and consequentially face downtime in collecting metrics.
> 
> User applications shouldn't use those labels. As far as I know, we treat shipped Prometheus instance as a system-one, integrated with components.
> 
> And making it backward compatible will result in duplicate targets and collected metrics :/ But still won't affect the monitoring downtime.

https://github.com/kinvolk/lokomotive/blob/master/docs/how-to-guides/monitoring-with-prometheus-operator.md#add-new-servicemonitors

https://github.com/kinvolk/lokomotive/blob/master/docs/how-to-guides/monitoring-with-prometheus-operator.md#add-custom-alerts-for-alertmanager

invidian replied:
https://github.com/kinvolk/lokomotive/pull/75 has a discussion about "system" Prometheus. Bummer that this is not really reflected in the docs.

ipochi replied:
@invidian created #1204 to track this in a separate issue.

surajssd reviewed (dismissed):
In step 2 of the PR description:

```
kubectl delete pvc/<PersistenceVolumeClaim name> -n monitoring
```

I think we know what the PVCs are called in monitoring namespace we should list them explicitly. If user has their own stuff here we just add a caveat that they handle them on their own.

In step 1 & step 3:

Since we already know the names of PVCs we can extract PV names from those PVCs, user does not have to bother with finding names.

The names we found in step 1 can be used in step 3.

ipochi commented:
> In step 2 of the PR description:
> 
> ```
> kubectl delete pvc/<PersistenceVolumeClaim name> -n monitoring
> ```
> 
> I think we know what the PVCs are called in monitoring namespace we should list them explicitly. If user has their own stuff here we just add a caveat that they handle them on their own.
> 
> In step 1 & step 3:
> 
> Since we already know the names of PVCs we can extract PV names from those PVCs, user does not have to bother with finding names.
> 
> The names we found in step 1 can be used in step 3.


For step 1, would this be legible ?
```
kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-alertmanager-0")].metadata.name}' | xargs kubectl patch pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
```

```
 kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="ddata-prometheus-prometheus-operator-prometheus-0")].metadata.name}' | xargs kubectl patch pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
```

and similar for step 3.

surajssd commented:
@ipochi yes for step1 it works fine. But then in step3 we won't have those PVs then above steps won't work. So we might need to save it in an intermediate vars.

ipochi commented:
> @ipochi yes for step1 it works fine. But then in step3 we won't have those PVs then above steps won't work. So we might need to save it in an intermediate vars.

@surajssd For step 3, we need to change the name of the PVCs.
```
 kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-kube-p-alertmanager-0")].metadata.name}' | xargs kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]'
```

invidian commented:
Why use `xargs` instead of `$()` ?

ipochi commented:
> Why use `xargs` instead of `$()` ?

That works too. My question is "Is it still legible ?"

invidian commented:
>That works too. My question is "Is it still legible ?"

Seems like a common pattern to me, so I don't mind.

ipochi commented:
@surajssd @invidian Updated release notes with the desired changes.

invidian commented:
I think changelog can be still simplified by using OR operator to just run one command instead of two.

ipochi commented:
@surajssd @invidian 

This step doesnt work if PVC name is used to get the PV.

3. Remove current `spec.claimRef` values to change the PV's status from Released to Available.
    ```console
    kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-kube-p-prometheus-0")].metadata.name}')

    kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-kube-p-alertmanager-0")].metadata.name}')
 ```

Not sure whats the problem here. Command works, but the status of second PVC changes to Bound state if the first command is applied and vice-versa.

Changing the `claimRef` with pv name works fine. Same problem with xargs

invidian commented:
>Not sure whats the problem here. Command works, but the status of second PVC changes to Bound state if the first command is applied and vice-versa.

I'll have a look.

invidian commented:
We could do the following to perform just one patch operation :smile: But it seems less readable to me:
```
kubectl patch pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}' $(for i in data-prometheus-prometheus-operator-prometheus-0 data-alertmanager-prometheus-operator-alertmanager-0; do kubectl get pv -o jsonpath="{.items[?(@.spec.claimRef.name=='$i')].metadata.name}"; echo -n " "; done; echo)
```

invidian commented:
For sure names in 3. are wrong:
```diff
-kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-kube-p-prometheus-0")].metadata.name}')
-kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-kube-p-alertmanager-0")].metadata.name}')
+kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-prometheus-0")].metadata.name}')
+kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-alertmanager-0")].metadata.name}')
```

invidian commented:
```
$ kgpv
+ kubectl get persistentvolume
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS      CLAIM   STORAGECLASS      REASON   AGE
pvc-932e66db-55d9-4ad4-a3ca-e674802b2c70   50Gi       RWO            Retain           Available           openebs-test-sc            33m
pvc-97b9059c-e482-4713-be41-1ef418b84193   50Gi       RWO            Retain           Available           openebs-test-sc            33m
```

After fixing names it seems to work for me :+1: 

I also run `lokoctl component delete` with `--delete-namespace`, to skip manual PVC removal.

invidian reviewed (changes requested):
Release notes needs fixing.

ipochi commented:
> Release notes needs fixing.

Will do.

ipochi commented:
> For sure names in 3. are wrong:
> 
> ```diff
> -kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-kube-p-prometheus-0")].metadata.name}')
> -kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-kube-p-alertmanager-0")].metadata.name}')
> +kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-prometheus-prometheus-operator-prometheus-0")].metadata.name}')
> +kubectl patch pv --type json -p='[{"op": "remove", "path": "/spec/claimRef"}]' $(kubectl get pv -o jsonpath='{.items[?(@.spec.claimRef.name=="data-alertmanager-prometheus-operator-alertmanager-0")].metadata.name}')
> ```

sorry I forgot to update the Release notes with the correct name, but I was using the names which you have and it didnt work.

ipochi commented:
> ```
> $ kgpv
> + kubectl get persistentvolume
> NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS      CLAIM   STORAGECLASS      REASON   AGE
> pvc-932e66db-55d9-4ad4-a3ca-e674802b2c70   50Gi       RWO            Retain           Available           openebs-test-sc            33m
> pvc-97b9059c-e482-4713-be41-1ef418b84193   50Gi       RWO            Retain           Available           openebs-test-sc            33m
> ```
> 
> After fixing names it seems to work for me 
> 
> I also run `lokoctl component delete` with `--delete-namespace`, to skip manual PVC removal.

Yes, we could do that step of `lokoctl component delete prometheus-operrator --delete-namespace`. However I didnt want to use that option as there could be other user created resources in that namespace.

ipochi commented:
@iaguis @invidian @johananl 
Can we get a consensus around this please ? Please provide your closing decisions so that I can act on the decision and make the required changes if any.

invidian commented:
@ipochi please rebase the PR to fix CI issues.

Regarding partial upgrades and backward-compatibility. Do we know how much work it would be to provide it and what are the implications?

johananl commented:
> @iaguis @invidian @johananl @surajssd
> Can we get a consensus around this please ? Please provide your closing arguments so that I can act on the decision and make the required changes if any.

@ipochi I'm not sure I personally can help by making a decision here. I'm also not sure we can - or should - always reach a consensus. I've shared my thoughts but you understand the issue better than I do. You are driving this task and I trust your judgement so as far as I'm concerned it's up to you :slightly_smiling_face:

I'll summarize my view in case it helps:

- I don't see an interruption to **monitoring** as "downtime".
- I see actual downtime (as in - dropped application traffic) as reasonable during upgrades as long as the process is deterministic and the downtime window is reasonable (i.e. minutes, not days) and the release notes clearly warn about it.
- If we are worried that the monitoring interruption could hide a service disruption caused by upgrading another component then maybe we should instruct the user to complete the monitoring upgrade before proceeding with the more risky upgrades.
- If we can explain why the label updates are necessary in the release notes and have a deterministic process (or even a script) for updating the labels then I don't see a problem with doing the "cutoff" in a single release.
- I'd also be fine with a "2-phase update" where release N makes a backward-compatible change and release N+1 does a cleanup as long as this doesn't mean leaving the cluster in a fragile state until the process is completed. Releases should be atomic.

If you were a user and you got this update from your k8s vendor, which option would you rather have?

invidian reviewed (changes requested):
Rebase required. Also new version of kube-prometheus-stack is now available.

ipochi commented:
@invidian @surajssd @iaguis @johananl 

Updated the Prometheus operator to 0.43.2 , rebased and additional duplicate resources so that the metrics wont be broken if:

- the user updates the cluster but not prometheus-operator
- the user updates cluster , prometheus-operator but not rook,contour or metallb.

Additional resources will become the main resources for scraping and the backward compatible ones will be removed in the next version of Lokomotive.

invidian commented:
It seems CI is broken.

ipochi commented:
> It seems CI is broken.

Yes. I'll look into when I return 🙂

ipochi commented:
@invidian @invidian @iaguis 
Build is passing. Feedback on metrics backward compatibility retained as mentioned [here](https://github.com/kinvolk/lokomotive/pull/1162#issuecomment-732179755).

Please review.

surajssd commented on assets/charts/components/contour/templates/alerts-for-prometheus-operator-0.43.2.yaml:
```diff
@@ -0,0 +1,33 @@
+# This resource is created to scrap contour metrics once Contour component is
```
`s/scrap/scrape/g`

surajssd replied:
Here and everywhere else in this commit.

surajssd commented on assets/charts/components/rook/templates/prometheus-ceph-v14-rules-for-prometheus-operator-0.43.2.yaml:
```diff
@@ -0,0 +1,262 @@
+# This resource is created to scrap rook metrics once Rook component is updated.
```
We should mention in the commit message when is the right time to remove it? Like after release of `0.6.0` (with an assumption that this PR will go as a part of 0.6.

surajssd replied:
In fact creating an issue is more beneficial in this case.

ipochi replied:
Created #1212

surajssd reviewed (commented):
small nits otherwise LGTM

surajssd reviewed (approved):
LGTM

invidian reviewed (approved):
LGTM

I added a comment in #1212, that maybe we can structure the commit here, that for the next release we only need to revert this commit without manual changes.

ipochi commented:
> LGTM
> 
> I added a comment in #1212, that maybe we can structure the commit here, that for the next release we only need to revert this commit without manual changes.

@invidian Just did that :)

Now the [commit](https://github.com/kinvolk/lokomotive/pull/1162/commits/b1eff74d4d92b07f6962aad6483035b5218cf733) should be removed from the branch once we are ready to roll out a newer version of Lokomotive > 0.6.0

invidian reviewed (approved):
LGTM

>@invidian Just did that :)

Thanks! It's looking good as far as I see :+1:
